Indicators/Covariance.cs

```diff
@@ -0,0 +1,197 @@
+        /// </summary>
+        /// <param name="name">The name of this indicator</param>
+        /// <param name="period">The period of this indicator</param>
+        /// <param name="targetSymbol">The target symbol of this indicator</param>
+        /// <param name="referenceSymbol">The reference symbol of this indicator</param>
+        /// <remarks>Constructor overload for backward compatibility.</remarks>
+        public Covariance(string name, int period, Symbol targetSymbol, Symbol referenceSymbol)
+            : this(name, targetSymbol, referenceSymbol, period)
+        {
+        }
+
+        /// <summary>
+        /// Computes the next value for this indicator from the given state.
+        /// 
+        /// As this indicator is receiving data points from two different symbols,
+        /// it's going to compute the next value when the amount of data points
+        /// of each of them is the same. Otherwise, it will return the last covariance
+        /// value computed
+        /// </summary>
+        /// <param name="input">The input value of this indicator on this time step.
+        /// It can be either from the target or the reference symbol</param>
+        /// <returns>The covariance value of the target used in relation with the reference</returns>
+        protected override decimal ComputeNextValue(IBaseDataBar input)
+        {
+            var inputSymbol = input.Symbol;
+            if (inputSymbol == _targetSymbol)
+            {
+                _targetDataPoints.Add(input.Value);
+            }
+            else if (inputSymbol == _referenceSymbol)
+            {
+                _referenceDataPoints.Add(input.Value);
+            }
+            else
+            {
+                throw new ArgumentException("The given symbol was not target or reference symbol");
+            }
+
+            if (_targetDataPoints.Samples == _referenceDataPoints.Samples && _referenceDataPoints.Count > 1)
+            {
+                ComputeCovariance();
+            }
+            
+            return _covariance;
+
+        }
+
+        /// <summary>
+        /// Computes the covariance value of the target in relation with the reference
+        /// using the target and reference returns
+        /// </summary>
+        private void ComputeCovariance()
+        {
+
+            if (!IsReady)
+            {
+                _covariance = 0m;
+                return;
+            }
+
+            decimal mean1 = 0;
+            decimal mean2 = 0;
+            decimal sumProduct = 0;
+
+            int windowSize = _targetDataPoints.Count;
+            for (int i = 0; i < windowSize; i++)
+            {
+                mean1 += _targetDataPoints[i] / windowSize;
+                mean2 += _referenceDataPoints[i] / windowSize;
+            }
+
+            for (int i = 0; i < windowSize; i++)
+            {
+                sumProduct += (_targetDataPoints[i] - mean1) * (_referenceDataPoints[i] - mean2);
+            }
```

Comment:
Take a look at the beta indicator which is calculating the covariance internally, as pointed by potential solution of the issue, I would suggest extracting that logic into this new covariance implementation and make the beta indicator the first user of it 👍